packages/core/src/conversation/ConversationService/ConversationService.ts

```diff
@@ -388,50 +424,173 @@ export class ConversationService extends TypedEventEmitter<Events> {
+    try {
+      await this.mlsService.joinByExternalCommit(() => this.apiClient.api.conversation.getGroupInfo(conversationId));
+    } catch (error) {
+      if (MLSService.isBrokenMLSConversationError(error)) {
+        this.logger.info(
+          "Failed to join MLS conversation via external commit because it's broken, resetting the conversation",
+          error,
+        );
+        if (!shouldRetry) {
+          this.logger.warn("Tried to join MLS conversation but it's still broken after reset", error);
+          throw error;
+        }
+        return this.handleBrokenMLSConversation(conversationId, (_newGroupId: string) =>
+          this.joinByExternalCommit(conversationId, false),
+        );
+      }
+      throw error;
+    }
   }
 
-  public async joinByExternalCommit(conversationId: QualifiedId) {
-    return this.mlsService.joinByExternalCommit(() => this.apiClient.api.conversation.getGroupInfo(conversationId));
+  private async resetMLSConversation(conversationId: QualifiedId): Promise<BaseCreateConversationResponse> {
+    this.logger.info(`Resetting MLS conversation with id ${conversationId.id}`);
+
+    // STEP 1: Fetch the conversation to retrieve the group ID & epoch
+    const conversation = await this.apiClient.api.conversation.getConversation(conversationId);
+    const {group_id: groupId, epoch} = conversation;
+
+    if (!groupId || !epoch) {
+      const errorMessage = 'Could not find group id or epoch for the conversation';
+      this.logger.error(errorMessage, {conversationId});
+      throw new Error(errorMessage);
+    }
+
+    // STEP 2: Request backend to reset the conversation
+    this.logger.info(`Requesting backend to reset the conversation (group_id: ${groupId}, epoch: ${String(epoch)})`);
+    await this.apiClient.api.conversation.resetMLSConversation({
+      epoch,
+      groupId,
+    });
+
+    // STEP 3: Re-establish the conversation by re-adding all members
+    this.logger.info(
+      `Re-establishing the conversation by re-adding all members (conversation_id: ${conversationId.id})`,
+    );
+    const clientId = this.apiClient.validatedClientId;
+    const {userId, domain} = this.apiClient;
+
+    if (!userId || !domain) {
+      const errorMessage = 'Could not find userId or domain of the self user';
+      this.logger.error(errorMessage, {conversationId});
+      throw new Error(errorMessage);
+    }
+
+    const selfUserQualifiedId = {id: userId, domain};
+
+    // STEP 4: Fetch the updated conversation data from backend to retrieve the new group ID
+    const updatedConversation = await this.apiClient.api.conversation.getConversation(conversationId);
+    const {group_id: newGroupId, members} = updatedConversation;
+
+    this.logger.info(`MLS conversation new group ID fetched from backend ${conversationId.id}`, {
+      newGroupId,
+    });
+    if (!newGroupId || !selfUserQualifiedId || !clientId) {
```

Comment:
How can `selfUserQualifiedId` possibly be undefined? You created a few lines above.